src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcProxyShimFactory.cs

```diff
@@ -0,0 +1,348 @@
+                resourceManagerIdentifier,
+                "System.Transactions.ResourceManager",
+                rmNotifyShim,
+                Guids.IID_IResourceManager_Guid,
+                out object? rm);
+
+            rmShim.ResourceManager = (IResourceManager)rm;
+        });
+
+        resourceManagerShim = rmShim;
+    }
+
+    public void Import(
+        byte[] cookie,
+        OutcomeEnlistment managedIdentifier,
+        out Guid transactionIdentifier,
+        out OletxTransactionIsolationLevel isolationLevel,
+        out TransactionShim transactionShim)
+    {
+        var txImport = (ITransactionImport)_transactionDispenser;
+        txImport.Import(Convert.ToUInt32(cookie.Length), cookie, Guids.IID_ITransaction_Guid, out object? tx);
+
+        SetupTransaction((ITransaction)tx, managedIdentifier, out transactionIdentifier, out isolationLevel, out transactionShim);
+    }
+
+    public void ReceiveTransaction(
+        byte[] propagationToken,
+        OutcomeEnlistment managedIdentifier,
+        out Guid transactionIdentifier,
+        out OletxTransactionIsolationLevel isolationLevel,
+        out TransactionShim transactionShim)
+    {
+        var receiver = GetCachedReceiver();
+
+        try
+        {
+            receiver.UnmarshalPropagationToken(
+                Convert.ToUInt32(propagationToken.Length),
+                propagationToken,
+                out ITransaction? tx);
+
+            SetupTransaction(tx, managedIdentifier, out transactionIdentifier, out isolationLevel, out transactionShim);
+        }
+        finally
+        {
+            ReturnCachedReceiver(receiver);
+        }
+    }
+
+    public void CreateTransactionShim(
+        IDtcTransaction transactionNative,
+        OutcomeEnlistment managedIdentifier,
+        out Guid transactionIdentifier,
+        out OletxTransactionIsolationLevel isolationLevel,
+        out TransactionShim transactionShim)
+    {
+        var cloner = (ITransactionCloner)transactionNative;
+        cloner.CloneWithCommitDisabled(out ITransaction transaction);
+
+        SetupTransaction(transaction, managedIdentifier, out transactionIdentifier, out isolationLevel, out transactionShim);
+    }
+
+    internal ITransactionExportFactory ExportFactory
+        => (ITransactionExportFactory)_transactionDispenser;
+
+    internal ITransactionVoterFactory2 VoterFactory
+        => (ITransactionVoterFactory2)_transactionDispenser;
+
+    public void GetNotification(
+        out object? managedIdentifier,
+        out ShimNotificationType shimNotificationType,
+        out bool isSinglePhase,
+        out bool abortingHint,
+        out bool releaseLock,
+        out byte[]? prepareInfo)
+    {
+        managedIdentifier = null;
+        shimNotificationType = ShimNotificationType.None;
+        isSinglePhase = false;
+        abortingHint = false;
+        releaseLock = false;
+        prepareInfo = null;
+
+        Monitor.Enter(_notificationLock);
```

Comment:
No.. I did change from a C++ linked list implementation to ConcurrentQueue.TryDequeue here, which AFAIK does not throw (and the rest of the code is just field accesses). I can add a try/catch to release just in case, though it doesn't seem necessary...

Let me know if you'd like to take a look at the original C++ code.

Reply:
Seems fine then.